### PR TITLE
Block public user from logining in

### PR DESCRIPTION
**Pull request by henryyang42**
Maybe it will work

### CODE REVIEW

File: users/views.py [hunk 1]
```diff
@@ -124,6 +125,7 @@ def user_login(request):
     if request.user.is_authenticated():
         return redirect(next_page)
     if request.method == 'POST':
+        deactivate_non_constant_public_users()
```
Comment (nickelcheng) [inline users/views.py]: I don't think it's necessary to do this everytime someone want to login. It causes a lot QQ.
Can you do that after the user is authenticated?

Comment (henryyang42) [inline users/views.py]: If the user is authenticated, it's hard to let the user logout.
It's easier that we check before we validate the form

Comment (henryyang42) [inline users/views.py]: Your suggestion is brilliant!
I will put it to the rightful place.

File: users/admin.py [hunk 2]
```diff
@@ -88,6 +89,13 @@ def __init__(self, *args, **kwargs):
         super(AuthenticationForm, self).__init__(*args, **kwargs)
         self.error_messages['inactive'] = 'This account is inactive. Check your email to activate the account!'
 
+    def confirm_login_allowed(self, user):
+        if is_public_user(user) and not attends_not_ended_contest(user):
```
Comment (henryyang42) [inline users/admin.py]: Check public user here.

Comment (nickelcheng) [inline users/admin.py]: It's good to override origin function!

**Comment by nickelcheng** [comment]
+1, going to merge.

**Pull request merged**